Feature/detect outside variables strands

Aayushdev18 opened the pull request:
Resolves #8172

## Changes

This PR adds detection for outside variable references in `p5.strands` uniform initializers, providing helpful error messages to users when they reference variables that aren't declared in the strand context.

**What was implemented:**
- Created `detectOutsideVariableReferences()` function in `src/strands/strands_transpiler.js` that:
  - Uses Acorn AST parsing to analyze strand code
  - Collects all variable declarations within the strand context
  - Finds uniform initializer functions (e.g., `uniformFloat('color', () => ...)`)
  - Extracts variables referenced inside uniform initializers
  - Checks if referenced variables are declared in the strand scope
  - Returns helpful error messages for any undeclared variables
- Integrated detection to run before transpilation in `src/strands/p5.strands.js`
- Added unit tests in `test/unit/strands/strands_transpiler.js` with test cases for:
  - Detecting undeclared variables (e.g., `mouseX`, `windowWidth`)
  - No errors when variables are properly declared
  - Multiple undeclared variables
  - Edge cases

**Approach:**
Following @davepagurek's feedback, this implementation:
- Runs as a separate analysis phase before transpilation
- Reuses the transpiler's Acorn-based structure for AST parsing
- Analyzes variable declarations to build a set of in-scope variables
- Detects when uniform initializer functions reference variables not in that set
- Provides specific, actionable error messages to users

This is more robust than checking for hardcoded variable names, as it analyzes the actual code structure and works with any user-defined variables.

## Screenshots

N/A (infrastructure change - no visual impact)

## PR Checklist

- [x] Code follows p5.js coding conventions and practices
- [x] Code is tested and all tests pass
- [x] Documentation is updated (if applicable)
- [x] Pull request targets the correct branch (dev-2.0)
- [x] Changes are minimal and focused on the issue
- [x] No breaking changes

Commit:
Add detection for outside variable references in p5.strands uniforms

- Create detectOutsideVariableReferences() function in strands_transpiler.js
- Uses Acorn AST parsing to collect declared variables from strand code
- Analyzes uniform initializer functions to find variable references
- Checks if referenced variables are declared in strand context
- Provides helpful error messages to users
- Integrates detection before transpilation in p5.strands.js
- Add unit tests for the detection logic

Addresses issue #8172 - Help users debug when accessing outside variables
in uniform initializers within p5.strands

Commit:
Fix outside variable detection: handle optional defaultValue and ancestors parameter

- Check for >= 2 arguments instead of > 0 for uniform functions
- Properly pass ancestors parameter to walk function
- Add comments clarifying uniform function signature

Commit:
Fix outside variable detection based on Dave's feedback

- Don't check uniform callbacks (they should allow outer scope)
- Check the main strand code for undeclared variables
- Use simpler two-pass approach: collect declarations, then check references
- Update tests to reflect correct behavior
- Handle function parameters and property access properly

Addresses Dave's comments about checking the wrong scenario

davepagurek commented on test/unit/strands/strands_transpiler.js:
```diff
@@ -0,0 +1,56 @@
+
+  test('should not error when variable is declared', function() {
+    const code = `
+      baseMaterialShader.modify(() => {
```
I think this is testing something different from what the function would actually receive. This looks like code that would be found in a p5 sketch, but just the contents of the `modify` callback is sent into the your function normally.

I think a good next step would be to try to do a full integration test by running `yarn dev:global` and editing the sketch in `preview/global`. If that works, you can also show the maintainers by building the library with `npm run build`, and then uploading `lib/p5.js` to a p5 web editor sketch so that others can run the same test. I suspect doing that will reveal some of the problems with the implementation more quickly than in review (but feel free to reach out if you run into issues while doing so!)

Commit:
Implement proper scope-aware variable detection

- Trace through ancestor chain to check if variable is in scope
- Only look for variable declarations in ancestor nodes
- Handles nested scopes properly (e.g. i declared in callback A, used in callback B)
- Simplify function parameter detection

Commit:
Simplify scope checking - ancestor chain naturally excludes sibling scopes

Commit:
Fix tests to match actual function input (just body, not wrapper)

Commit:
Fix outside variable detection in p5.strands

- Fixed function parameter detection by comparing parameter names instead of AST nodes
- Added comprehensive built-in function list to ignore list
- Simplified scope detection logic using two-pass AST traversal
- All unit tests now passing
- Error detection working correctly in integration tests

Aayushdev18 committed:
Merge branch 'dev-2.0' into feature/detect-outside-variables-strands

Aayushdev18 committed:
Merge branch 'dev-2.0' into feature/detect-outside-variables-strands

Aayushdev18 committed:
Merge branch 'dev-2.0' into feature/detect-outside-variables-strands

davepagurek commented on src/strands/strands_transpiler.js:
```diff
@@ -920,7 +920,100 @@ const ASTCallbacks = {
+        const ignoreNames = [
+          '__p5', 'p5', 'window', 'global', 'undefined', 'null', 'this', 'arguments',
+          // p5.strands built-in functions
+          'getWorldPosition', 'getWorldNormal', 'getWorldTangent', 'getWorldBinormal',
```
This approach is currently hard-coding these identifiers, but they can be almost anything (a custom shader can create its own hooks for other users), so we'll need a different approach. A number of the identifiers in this list don't appear anywhere in any default shaders either, such as `getWorldBinormal`, and feel like AI hallucinations. Given the complex nature of this task and the system it's interacting with, I think we really need to have a deeper understanding of what the approach here should be before proceeding with writing code, otherwise it will take a significant amount of back-and-forth in review. Is that something you're up for? If so, we can discuss an approach to how to tackle this (and feel free to reach out in the p5 discord, I'm happy to explain in a less asynchronous way there!) If not, also no worries -- this is a complex issue, and there are others that you can tackle, and we can open this one up to other contributors.
